include/dslash.h

```diff
@@ -323,23 +394,39 @@ namespace quda
 #endif
     }
 
+    void setShmem(int shmem_)
+    {
+      arg.shmem = shmem_;
+      setUberTuning(arg.shmem & 64);
+    }
+
     void setPack(bool pack, MemoryLocation location)
     {
-      arg.setPack(pack);
-      if (!pack) return;
+      if (!pack) {
+        arg.setPack(pack, packBuffer);
+        return;
+      }
 
       for (int dim = 0; dim < 4; dim++) {
         for (int dir = 0; dir < 2; dir++) {
           if ((location & Remote) && comm_peer2peer_enabled(dir, dim)) { // pack to p2p remote
             packBuffer[2 * dim + dir] = static_cast<char *>(in.remoteFace_d(dir, dim)) + in.GhostOffset(dim, 1 - dir);
           } else if (location & Host && !comm_peer2peer_enabled(dir, dim)) { // pack to cpu memory
             packBuffer[2 * dim + dir] = in.myFace_hd(dir, dim);
+          } else if (location & Shmem) {
+            packBuffer[2 * dim + dir] = in.remoteFace_d(dir, dim) ?
```

Comment:
What the logic for querying `remoteFace_d` here? Is it to support `--partition`?

Reply:
if `remoteFace_d` then we can directly pack to the remote. The packing destination is `packBuffer[2* dim + dir]`. If not the packing writes to `in.myFace_d(dir, dim);` and then we use `nvshmem_putmem` to get the data to `static_cast<char *>(in.remoteFace_r()) + in.GhostOffset(dim, 1 - dir);` (copy from `packBuffer[2* dim + dir]` to `packBuffer[2*QUDA_MAX_Dim + 2* dim + dir]`